File: SqlViewGenerator/JsonModel/Agregators/Conditions/ConditionLink.cs
﻿namespace SqlViewGenerator.JsonModel.Agregators.Conditions;

public class ConditionLink : IVisitable
{
    public enum LinkRelation
    {
        And,
        Or,
    }

    public ConditionLink(LinkRelation relation, JoinCondition condition)
    {
        this.Relation = relation;
        this.Condition = condition;
    }

    public LinkRelation Relation { get; }

    public JoinCondition Condition { get; }

    public void Accept(IVisitor visitor)
    {
        visitor.Visit(this);
    }
}

File: src/Modules/DataIntegration/MapperPages/Mapper/src/mappingModel/aggregators/conditions/conditionLink.ts
import { ReferenceHolder } from "../../referenceHolder";
import { MappingVisitor } from "../../mappingVisitor";
import { Visitable } from "../../converting/visitable";
import type { JoinCondition } from "./joinCondition";

export enum LinkRelation {
    and = "and",
    or = "or"
}

export class ConditionLink implements Visitable, ReferenceHolder {
    /**
     * Inializes a new instance of ConditionLink
     * @param relation The relation between the conditions.
     * @param condition The next condition in the chain.
     * Should be of type JoinCondition, but we can't use it because of the circular dependency.
     */
    public constructor(
        public relation : LinkRelation,
        public condition : JoinCondition) {
    }
    removeReferences(): void {
        this.condition.removeReferences();
    }

    accept(visitor: MappingVisitor): void {
        visitor.visitConditionLink(this);
    }

    createBackwardConnections(): void {
        this.condition.createBackwardConnections();
    }
}